### ISSUE TITLE
feat(gp): ARD (per-dimension) lengthscales for RBF, Matern, RationalQuadratic

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#198]
## Summary

`pyrox_gp` kernels currently accept only a **scalar** lengthscale.
`_src/kernels.py` annotates every one as `Float[Array, ""]`, and every
`_ParameterizedKernel` subclass registers `jnp.asarray(self.init_lengthscale)`
from a `float` field. There is no way to fit a per-input-dimension lengthscale.

Add opt-in ARD to the three kernels where a per-dimension lengthscale is
well-defined: `RBF`, `Matern`, `RationalQuadratic`.

## Motivation

ARD is table stakes for any GP on `D > 1` inputs — it is the mechanism by which
a GP does automatic input relevance selection, and its absence is a real
capability gap independent of any other work in flight. It also unblocks
per-latent ARD lengthscales for multi-output and latent-factor models
(issue [C]).

There is a second, quieter win. The natural way to write ARD is an explicit
`(N1, N2, D)` difference tensor. Scaling the *inputs* before the
`‖x‖² + ‖x'‖² − 2x·x'` expansion gets ARD while keeping the current `O(N²)`
memory. For `N = 2000, D = 50` that is **32 MB instead of 1.6 GB**.

## Scope — which kernels, and why not the others

| Kernel | ARD? | Reason |
|---|---|---|
| `RBF` | ✅ | `lengthscale` scales the distance directly |
| `Matern` (`nu` ∈ {0.5, 1.5, 2.5}) | ✅ | same |
| `RationalQuadratic` | ✅ | same |
| `Periodic` | ❌ | `lengthscale` divides `sin²(πr/p)`, not the distance. `period` is the length-like parameter and is scalar by construction. Out of scope. |
| `Cosine`, `White`, `Constant`, `Linear`, `Polynomial` | ❌ | no lengthscale parameter |

### 1. `_src/kernels.py` — scale inside the distance primitive

```python
def _pairwise_sq_dist(
    X1: Float[Array, "N1 D"],
    X2: Float[Array, "N2 D"],
    lengthscale: Float[Array, ""] | Float[Array, " D"] = 1.0,
) -> Float[Array, "N1 N2"]:
    """Squared Euclidean distance matrix, optionally lengthscale-scaled.

    Scaling the inputs *before* the
    $\\lVert x \\rVert^2 + \\lVert x' \\rVert^2 - 2 x^\\top x'$ expansion
    supports a per-dimension lengthscale while keeping every intermediate at
    ``(N1,)``, ``(N2,)``, or ``(N1, N2)`` — no ``(N1, N2, D)`` broadcast
    tensor is ever built. Clipped at zero to absorb float cancellation on
    near-identical points.

    Args:
        X1: ``(N1, D)`` inputs.
        X2: ``(N2, D)`` inputs.
        lengthscale: Scalar (isotropic) or ``(D,)`` (ARD) lengthscale.
            Broadcasts against the trailing input axis in both cases.

    Returns:
        ``(N1, N2)`` scaled squared distances.
    """
    X1 = X1 / lengthscale
    X2 = X2 / lengthscale
    n1 = einx.dot("n1 d, n1 d -> n1", X1, X1)
    n2 = einx.dot("n2 d, n2 d -> n2", X2, X2)
    cross = einx.dot("n1 d, n2 d -> n1 n2", X1, X2)
    norm_sum = einx.add("n1, n2 -> n1 n2", n1, n2)
    return jnp.clip(norm_sum - 2.0 * cross, min=0.0)
```

Then update the three call sites to pass the lengthscale in and **stop dividing
by it again**:

```python
# rbf_kernel
sq = _pairwise_sq_dist(X1, X2, lengthscale)
return variance * jnp.exp(-0.5 * sq)              # was: / (lengthscale ** 2)

# matern_kernel
sq = _pairwise_sq_dist(X1, X2, lengthscale)
r = jnp.sqrt(jnp.clip(sq, min=1e-30))             # was: ... / lengthscale
...

# rational_quadratic_kernel
sq = _pairwise_sq_dist(X1, X2, lengthscale)
return variance * (1.0 + sq / (2.0 * alpha)) ** (-alpha)
```

> ⚠️ **The most likely bug in this change is double-scaling.** Each of the three
> kernels currently divides by `lengthscale` *after* calling
> `_pairwise_sq_dist`. Once the scaling moves into the primitive, those
> divisions must be removed. The back-compatibility test below is what catches
> this — it will fail loudly with `ell ≠ 1`.

Leave `periodic_kernel` and `cosine_kernel` calling `_pairwise_sq_dist(X1, X2)`
with the default — their `r` is a raw distance fed to a periodic function.

Also widen the `lengthscale` annotation on the three public kernel functions to
`Float[Array, ""] | Float[Array, " D"]` and update their `Args:` docstrings.

### 2. `_kernels.py` — opt-in via an `input_dim` field

```python
class RBF(_ParameterizedKernel):
    """Radial basis function (squared exponential) kernel."""

    pyrox_name: str = "RBF"
    init_variance: float = 1.0
    init_lengthscale: float = 1.0
    input_dim: int | None = None

    def setup(self) -> None:
        self.register_param(
            "variance",
            jnp.asarray(self.init_variance),
            constraint=dist.constraints.positive,
        )
        lengthscale = (
            jnp.asarray(self.init_lengthscale)
            if self.input_dim is None
            else jnp.full((self.input_dim,), self.init_lengthscale)
        )
        self.register_param(
            "lengthscale",
            lengthscale,
            constraint=dist.constraints.positive,
        )
```

Identical change for `Matern` and `RationalQuadratic`. `input_dim=None` keeps
the current scalar behaviour, so this is fully backwards compatible and no
existing call site changes.

Document `input_dim` in each class docstring: *"Set to the input dimension `D`
to fit a separate lengthscale per input dimension (ARD). Leave as `None` for a
single isotropic lengthscale."*

`input_dim` is a shape-determining static field, consistent with how `nu` and
`degree` are already handled (`_kernels.py:80-84`, `:239`).

## Usage after the change

```python
from pyrox_gp import RBF, Matern, GPPrior

k_iso = RBF(init_lengthscale=0.5)                     # unchanged
k_ard = RBF(input_dim=3, init_lengthscale=0.5)        # per-dimension
k_ard_matern = Matern(input_dim=3, nu=1.5)

prior = GPPrior(kernel=k_ard, X=X)                    # X: (N, 3)
```

Per-latent ARD, distinct `pyrox_name` per instance so the sites do not collide
(see `_validate_kernel_scopes_unique` in `_multi_output.py:49`):

```python
kernels = tuple(
    RBF(pyrox_name=f"RBF_q{q}", input_dim=3) for q in range(Q)
)
```

## Acceptance criteria

Extend `packages/pyrox-gp/tests/gp/test_kernels.py` and
`test_kernel_classes.py`. Enable x64 for the tight-tolerance cases.

- [ ] **Back-compatibility (catches double-scaling).** For each of RBF, Matern
      (all three `nu`), RationalQuadratic, with a **non-unit** scalar
      lengthscale (e.g. `0.7`), the new implementation matches the pre-change
      values to `1e-12`. Hard-code the expected Gram from the current
      implementation, or assert against a hand-written reference — do not
      compare the new code to itself.
- [ ] **ARD correctness.** With `lengthscale = (D,)`, matches an explicit
      reference built from a `(N1, N2, D)` difference tensor,
      `sum(((X1[:, None] - X2[None]) / ell) ** 2, -1)`. Tolerance `1e-12`.
- [ ] **Degenerate ARD.** `jnp.full((D,), c)` gives bitwise-identical results to
      scalar `c`.
- [ ] **Gradients.** `jax.grad` w.r.t. a `(D,)` lengthscale is finite and has
      shape `(D,)`, **including on a self-Gram** where the diagonal has `r = 0`.
- [ ] **`vmap` over per-latent lengthscales.** `jax.vmap` of a kernel over a
      `(Q, D)` lengthscale array produces `(Q, N, N)` finite Grams under `jit`.
- [ ] **Class-level opt-in.** `RBF(input_dim=3).get_param("lengthscale")` has
      shape `(3,)`; `RBF().get_param("lengthscale")` stays scalar.
- [ ] **Priors still attach.** `set_prior("lengthscale", ...)` works on an ARD
      kernel and produces a correctly-shaped NumPyro site (check event shape —
      a `(D,)` lengthscale needs `.to_event(1)` semantics from the caller).
- [ ] **Excluded kernels unchanged.** `Periodic` and `Cosine` produce
      bit-identical output to before.
- [ ] **Memory.** An ARD Gram at `N = 2000, D = 50` completes without building
      an `(N, N, D)` intermediate.

## Validation evidence

The refactor above was prototyped and executed:

| Check | Result |
|---|---|
| RBF scalar `ell=0.7`, old vs new | `1.1e-15` |
| Matern-5/2 scalar `ell=0.7`, old vs new | `7.8e-16` |
| RBF ARD vs explicit `(N1,N2,D)` reference | `1.3e-15` |
| ARD with constant vector vs scalar | `0.0` exactly |
| `grad` w.r.t. `(D,)` lengthscale on self-Gram | finite, shape `(4,)` |
| `jit(vmap(...))` over `(6, 4)` lengthscales | `(6, 17, 17)`, finite |
| `N=2000, D=50` ARD Gram | 32 MB vs 1.6 GB for the tensor form |

## Note on a non-issue

An earlier review flagged possible `NaN` gradients from `sqrt` at `r = 0`.
**This is already handled** — `matern_kernel`, `periodic_kernel`, and
`cosine_kernel` all use `jnp.sqrt(jnp.clip(sq, min=1e-30))` with an explicit
comment (`_src/kernels.py:107`). No change needed; preserve it.

## Out of scope

- ARD for `Periodic` (see scope table)
- A separate ARD variant of `Linear` / `Polynomial`
- Changing the default from isotropic to ARD
- `gaussx`-side stable kernel construction (`stable_rbf_kernel`) — separate repo

## Before reporting done

```bash
uv run pytest -q
uv run --group lint ruff check .
uv run --group lint ruff format --check .
uv run --group typecheck ty check packages/pyrox-gp/src/pyrox_gp
```
